state/in-memory/in_memory.go

```diff
@@ -174,78 +174,72 @@ func (store *inMemoryStore) BulkDelete(ctx context.Context, req []state.DeleteRe
 }
 
 func (store *inMemoryStore) Get(ctx context.Context, req *state.GetRequest) (*state.GetResponse, error) {
-	item := store.doGetWithReadLock(ctx, req.Key)
-	if item != nil && isExpired(item) {
-		item = store.doGetWithWriteLock(ctx, req.Key)
+	store.lock.RLock()
+	item := store.items[req.Key]
+	store.lock.RUnlock()
+	if item != nil && item.isExpired() {
+		store.lock.Lock()
+		item = store.getAndExpire(req.Key)
+		store.lock.Unlock()
 	}
 
 	if item == nil {
-		return &state.GetResponse{Data: nil, ETag: nil}, nil
+		return &state.GetResponse{}, nil
 	}
 
-	data := item.data
-	if item.isBinary {
-		var (
-			s   string
-			err error
-		)
-
-		if err = jsoniter.Unmarshal(data, &s); err != nil {
-			return nil, err
-		}
+	return &state.GetResponse{Data: item.data, ETag: item.etag}, nil
+}
 
-		data, err = base64.StdEncoding.DecodeString(s)
-		if err != nil {
-			return nil, err
-		}
+func (store *inMemoryStore) BulkGet(ctx context.Context, req []state.GetRequest, _ state.BulkGetOpts) ([]state.BulkGetResponse, error) {
+	res := make([]state.BulkGetResponse, len(req))
+	if len(req) == 0 {
+		return res, nil
 	}
 
-	return &state.GetResponse{Data: data, ETag: item.etag}, nil
-}
-
-func (store *inMemoryStore) doGetWithReadLock(ctx context.Context, key string) *inMemStateStoreItem {
+	// While working in bulk, we won't delete expired records we may encounter; we'll just let them stay until GC picks them up
 	store.lock.RLock()
 	defer store.lock.RUnlock()
 
-	return store.items[key]
+	n := 0
+	for i, r := range req {
+		item := store.items[r.Key]
+		if item != nil && !item.isExpired() {
+			res[i] = state.BulkGetResponse{
+				Key:  r.Key,
+				Data: item.data,
+				ETag: item.etag,
+			}
+			n++
+		}
+	}
+
+	return res[:n], nil
 }
 
-func (store *inMemoryStore) doGetWithWriteLock(ctx context.Context, key string) *inMemStateStoreItem {
-	store.lock.Lock()
-	defer store.lock.Unlock()
+func (store *inMemoryStore) getAndExpire(key string) *inMemStateStoreItem {
 	// get item and check expired again to avoid if item changed between we got this write-lock
 	item := store.items[key]
 	if item == nil {
 		return nil
 	}
-	if isExpired(item) {
-		store.doDelete(ctx, key)
+	if item.isExpired() {
+		delete(store.items, key)
 		return nil
 	}
 	return item
 }
 
-func isExpired(item *inMemStateStoreItem) bool {
-	if item == nil || item.expire == nil {
-		return false
-	}
-	return time.Now().UnixMilli() > *item.expire
-}
-
-func (store *inMemoryStore) BulkGet(ctx context.Context, req []state.GetRequest) (bool, []state.BulkGetResponse, error) {
-	return false, nil, nil
-}
-
-func (store *inMemoryStore) marshal(v any) (bt []byte, isBinary bool, err error) {
+func (store *inMemoryStore) marshal(v any) (bt []byte, err error) {
```

Comment:
This code has been simplified since we don't need to marshal the data to JSON for then un-marshalling it when retrieving it. The data is kept in memory so that was an extra step.